src/machine.rs

```diff
@@ -430,6 +430,41 @@ impl<'mir, 'tcx> Evaluator<'mir, 'tcx> {
                 this.write_scalar(Scalar::from_u8(0), &place.into())?;
                 Self::add_extern_static(this, "_tls_used", place.ptr);
             }
+            "android" => {
+                // "environ"
+                Self::add_extern_static(
+                    this,
+                    "environ",
+                    this.machine.env_vars.environ.unwrap().ptr,
+                );
+                // A couple zero-initialized pointer-sized extern statics.
+                // Most of them are for weak symbols, which we all set to null (indicating that the
+                // symbol is not supported, and triggering fallback code which ends up calling a
+                // syscall that we do support).
+                for name in &["__cxa_thread_atexit_impl", "getrandom", "statx"] {
+                    let layout = this.machine.layouts.usize;
+                    let place = this.allocate(layout, MiriMemoryKind::ExternStatic.into())?;
+                    this.write_scalar(Scalar::from_machine_usize(0, this), &place.into())?;
+                    Self::add_extern_static(this, name, place.ptr);
+                }
+                for symbol_name in &["signal", "bsd_signal"] {
+                    let layout = this.machine.layouts.mut_raw_ptr;
```

Comment:
Just for the sake of documentation this is what caused the tests to fail after rebasing.

Previously we used `this.machine.layouts.usize` but strict provenance refactoring now expects us to use `mut_raw_ptr`. Using `usize` raised the following panic:

```
thread 'rustc' panicked at 'Miri initialization error: type validation failed: encountered pointer to 0x2006c[alloc6]<6>, but expected plain (non-pointer) bytes', src/eval.rs:321:13
```

The fullish stack trace is:

```
An error occurred in miri:
   0: std::backtrace::Backtrace::create
   1: std::backtrace::Backtrace::force_capture
   2: <rustc_middle::mir::interpret::error::InterpErrorInfo as core::convert::From<rustc_middle::mir::interpret::error::InterpError>>::from
   3: rustc_const_eval::interpret::validity::ValidityVisitor<M>::try_visit_primitive
   4: <rustc_const_eval::interpret::validity::ValidityVisitor<M> as rustc_const_eval::interpret::visitor::ValueVisitor<M>>::visit_value
             at /rustc/e71440575c930dcecac288b7c3536410d688b351/compiler/rustc_const_eval/src/interpret/validity.rs:807:12
   5: rustc_const_eval::interpret::validity::<impl rustc_const_eval::interpret::eval_context::InterpCx<M>>::validate_operand_internal
             at /rustc/e71440575c930dcecac288b7c3536410d688b351/compiler/rustc_const_eval/src/interpret/validity.rs:982:15
   6: rustc_const_eval::interpret::validity::<impl rustc_const_eval::interpret::eval_context::InterpCx<M>>::validate_operand
             at /rustc/e71440575c930dcecac288b7c3536410d688b351/compiler/rustc_const_eval/src/interpret/validity.rs:1023:9
   7: rustc_const_eval::interpret::place::<impl rustc_const_eval::interpret::eval_context::InterpCx<M>>::write_immediate
             at /rustc/e71440575c930dcecac288b7c3536410d688b351/compiler/rustc_const_eval/src/interpret/place.rs:675:13
   8: rustc_const_eval::interpret::place::<impl rustc_const_eval::interpret::eval_context::InterpCx<M>>::write_scalar
             at /rustc/e71440575c930dcecac288b7c3536410d688b351/compiler/rustc_const_eval/src/interpret/place.rs:688:9
   9: miri::machine::Evaluator::init_extern_statics
             at ./src/machine.rs:463:21
  10: miri::machine::Evaluator::late_init
             at ./src/machine.rs:389:9
  11: miri::eval::create_ecx
             at ./src/eval.rs:180:5
  12: miri::eval::eval_entry
             at ./src/eval.rs:317:38
  13: <miri::MiriCompilerCalls as rustc_driver::Callbacks>::after_analysis::{{closure}}
             at ./src/bin/miri.rs:78:40
  14: rustc_interface::passes::QueryContext::enter::{{closure}}
             at /rustc/e71440575c930dcecac288b7c3536410d688b351/compiler/rustc_interface/src/passes.rs:813:42
  15: rustc_middle::ty::context::tls::enter_context::{{closure}}
             at /rustc/e71440575c930dcecac288b7c3536410d688b351/compiler/rustc_middle/src/ty/context.rs:1873:50
  16: rustc_middle::ty::context::tls::set_tlv
             at /rustc/e71440575c930dcecac288b7c3536410d688b351/compiler/rustc_middle/src/ty/context.rs:1857:9
  17: rustc_middle::ty::context::tls::enter_context
             at /rustc/e71440575c930dcecac288b7c3536410d688b351/compiler/rustc_middle/src/ty/context.rs:1873:9
  18: rustc_interface::passes::QueryContext::enter
             at /rustc/e71440575c930dcecac288b7c3536410d688b351/compiler/rustc_interface/src/passes.rs:813:9
```

Reply:
If these are actually pointers and we are storing things with provenance there, then indeed this sounds like the more accurate type to use. :)